access/ms-teams/bot.go

```diff
@@ -0,0 +1,237 @@
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/trace"
+)
+
+// UserData represents cached data for a user
+type UserData struct {
+	// User ID user id
+	ID string
+	// App app installation for user
+	App msapi.InstalledApp
+	// Chat chat for user
+	Chat msapi.Chat
+}
+
+// Bot represents the facade to MS teams API
+type Bot struct {
+	// Config MS API configuration
+	msapi.Config
+	// teamsApp represents MS Teams app installed for an org
+	teamsApp *msapi.TeamsApp
+	// graphClient represents MS API Graph client
+	graphClient *msapi.GraphClient
+	// botClient represents MS Bot Framework client
+	botClient *msapi.BotFrameworkClient
+	// mu users access mutex
+	mu *sync.RWMutex
+	// apps represents the cache of apps
+	users map[string]UserData
+	// webProxyURL represents Web UI address, if enabled
+	webProxyURL *url.URL
+	// clusterName cluster name
+	clusterName string
+}
+
+// NewBot creates new bot struct
+func NewBot(c msapi.Config, clusterName, webProxyAddr string) (*Bot, error) {
+	var (
+		webProxyURL *url.URL
+		err         error
+	)
+
+	if webProxyAddr != "" {
+		webProxyURL, err = lib.AddrToURL(webProxyAddr)
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+	}
+
+	bot := &Bot{
+		Config:      c,
+		graphClient: msapi.NewGraphClient(c),
+		botClient:   msapi.NewBotFrameworkClient(c),
+		users:       make(map[string]UserData),
+		webProxyURL: webProxyURL,
+		clusterName: clusterName,
+		mu:          &sync.RWMutex{},
+	}
+
+	return bot, nil
+}
+
+// GetTeamsApp finds the application in org store and caches it in a bot instance
+func (b *Bot) GetTeamsApp(ctx context.Context) (*msapi.TeamsApp, error) {
+	teamsApp, err := b.graphClient.GetTeamsApp(ctx, b.Config.TeamsAppID)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	b.teamsApp = teamsApp
+	return b.teamsApp, nil
+}
+
+// GetUserIDByEmail gets a user ID by email. NotFoundError if not found.
+func (b Bot) GetUserIDByEmail(ctx context.Context, email string) (string, error) {
+	user, err := b.graphClient.GetUserByEmail(ctx, email)
+	if trace.IsNotFound(err) {
+		return "", trace.Wrap(err, "try user id instead")
+	} else if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	return user.ID, nil
+}
+
+// UserExists return true if a user exists. Returns NotFoundError if not found.
+func (b Bot) UserExists(ctx context.Context, id string) error {
+	_, err := b.graphClient.GetUserByID(ctx, id)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	return nil
+}
+
+// FetchUser fetches app id for user, installs app for a user if missing, fetches chat id and saves
+// everything to cache. This method is used for priming the cache. Returns trace.NotFound if a
+// user was not found.
+func (b Bot) FetchUser(ctx context.Context, userIDOrEmail string) (*UserData, error) {
+	if b.teamsApp == nil {
+		return nil, trace.Errorf("Bot is not configured, run GetTeamsApp first")
+	}
+
+	b.mu.RLock()
+	d, ok := b.users[userIDOrEmail]
```

Comment:
Do we need to restart the plugin if we want a newly created user to receive notifications?